### DISCUSSION TITLE
Background script steals divs/notifications

### DISCUSSION THREAD

**Discussion by hakubo** [johnlindquist/kit#493] [Errors] [answered]
I tried to create a simple script to time when I sit/stand during work. But unfortunately it appears it's not easyto show UI from background script in a way that do not interfere with main app. When I use 'notify' - it appears to close the script as well

https://user-images.githubusercontent.com/1018759/138547233-9ff660c7-1b08-4a69-8f0f-6c2e86807253.mp4

Example script: 

```js
// Background: auto

/** @type {import("@johnlindquist/kit")} */
const sitFor = 3000;
const standFor = 3000;

function startStand() {
  setTimeout(() => {
    div("time to sit!");

    startSit();
  }, standFor);
}

function startSit() {
  setTimeout(() => {
    div("time to stand!");

    startStand();
  }, sitFor);
}

startSit();

```

**Comment by johnlindquist** [comment, marked as answer]
One rule I put in place early on: 1 UI = 1 script. So starting a new script always cancels a currently running script in the app.

The reason: I don't want Script Kit to accidentally create a ton of runaway processes that could end up crashing the machine (which could be really easy to do with background/watchers/incoming events/etc).

## Solution

You should be able to run a script that uses `notify` in the terminal. Running a script in the terminal doesn't open a channel to the app.

## Future Solution

I still plan on adding a "widget" system to spawn multiple UIs that each bind to a specific script (imagine spawning multiple editors, etc). But that's sometime next year.


## Notes on the script:

You should be able to take advantage of `await` for timed loops:

```js
while(true){
    await wait(3000)
    notify(`stand`)
    await wait(3000)
    notify(`sit`)
}
```

**Comment by hakubo** [comment]
Ah got it! Thanks for explaining. 

but it appears that when you call notify from background task - UI of Kit is disappearing.


https://user-images.githubusercontent.com/1018759/138568140-4ce9fc49-2bae-4e2a-a2d0-a6d95f8acf17.mp4

// Thanks for the idea with while + await. I'll share the script once I polish it :)

**Reply by johnlindquist** [reply]
Hmm, that might be a focus issue. The UI is meant to close (and cancel script) when losing focus (which is usually a "click outside"). I'll dig into it a bit. 

"ignoreBlur" on an arg is usually the solution, but I'll need to check on div, etc. If that's not it, I'll dig into it

**Comment by hakubo** [comment]
Idea: clicking on kit icon in top bar - could show some information from background scripts (something like: https://xbarapp.com/)

**Reply by johnlindquist** [reply]
I like the idea of using the tray as an info panel. I'll have to mock out some ideas.

I could also dynamically add a tab to the main UI if it detects any background/cron/etc tasks are running.

Did you know about the Schedule/cron metadata. People are using that for scripts like these

**Reply by hakubo** [reply]
I know but then I'd have to keep the info in a DB. But maybe that's a better solution. I'll dig into it :)

**Comment by johnlindquist** [comment]
I finally tracked this down. I'll get the fix in today.

## Explanation

`notify` was sending a `HIDE_APP` message to the app, even though you were using a `Background` process type. Now the `HIDE_APP` filters to only trigger when it receives from a `Prompt` process type.

Based on the metadata you use, it sends along the ProcessType since the app handles each a little diffrently:

```js
export declare enum ProcessType {
    App = "App",
    Background = "Background",
    Prompt = "Prompt",
    Schedule = "Schedule",
    System = "System",
    Watch = "Watch"
}
```
